tests/sql/sla_test.go

```diff
@@ -47,6 +47,22 @@ func TestSla(t *testing.T) {
 		Start:    1000,
 		End:      2000,
 		Expected: 60.0,
+	}, {
+		Name: "MultipleStateChangesOddNumbers",
+		// Some flapping again, also test that calculations are rounded correctly including decimal places.
+		Events: []SlaHistoryEvent{
+			&State{Time: 1000, State: 2, PreviousState: 99}, // -2.3%
+			&State{Time: 1023, State: 0, PreviousState: 2},
+			&State{Time: 1100, State: 2, PreviousState: 0}, // -14.2%
+			&State{Time: 1242, State: 0, PreviousState: 2},
+			&State{Time: 1300, State: 2, PreviousState: 0}, // -0.7%
+			&State{Time: 1307, State: 0, PreviousState: 2},
+			&State{Time: 1400, State: 2, PreviousState: 0}, // -26.6%
+			&State{Time: 1666, State: 0, PreviousState: 2},
+		},
+		Start:    1000,
+		End:      2000,
+		Expected: 56.2,
```

Comment:
A test that uses the full number of decimal places available would be nice. Maybe even by having a repeating decimal as the result which is truncated by the stored function.

Reply:
Good catch. By doing so, I realized that a `decimal` cast alone wasn't enough to satisfy the `RETURNS decimal(7, 4)` specification.

Reply:
I just removed the final `decimal(7, 4)` cast, as shown in the diff. Without the tuple of precision and scale, there were more decimal places.

```diff
diff --git a/schema/pgsql/schema.sql b/schema/pgsql/schema.sql
index 1302103..84f06fa 100644
--- a/schema/pgsql/schema.sql
+++ b/schema/pgsql/schema.sql
@@ -166,7 +166,7 @@ BEGIN
     END IF;
   END LOOP;

-  RETURN (100 * (total_time - problem_time)::decimal / total_time)::decimal(7, 4);
+  RETURN (100 * (total_time - problem_time)::decimal / total_time);
 END;
 $$;
```

```
=== RUN   TestSla
=== RUN   TestSla/MultipleStateChangesDecimalsOddNumbers
=== RUN   TestSla/MultipleStateChangesDecimalsOddNumbers/Host
=== RUN   TestSla/MultipleStateChangesDecimalsOddNumbers/Service
=== RUN   TestSla/MultipleStateChangesDecimalsFractionOneThird
=== RUN   TestSla/MultipleStateChangesDecimalsFractionOneThird/Host
    sla_test.go:310:
                Error Trace:    /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:310
                                                        /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:274
                Error:          Not equal:
                                expected: 66.6667
                                actual  : 66.66666666666667
                Test:           TestSla/MultipleStateChangesDecimalsFractionOneThird/Host
                Messages:       unexpected SLA value
=== RUN   TestSla/MultipleStateChangesDecimalsFractionOneThird/Service
    sla_test.go:310:
                Error Trace:    /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:310
                                                        /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:277
                Error:          Not equal:
                                expected: 66.6667
                                actual  : 66.66666666666667
                Test:           TestSla/MultipleStateChangesDecimalsFractionOneThird/Service
                Messages:       unexpected SLA value
=== RUN   TestSla/MultipleStateChangesDecimalsFractionSeventhPart
=== RUN   TestSla/MultipleStateChangesDecimalsFractionSeventhPart/Host
    sla_test.go:310:
                Error Trace:    /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:310
                                                        /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:274
                Error:          Not equal:
                                expected: 14.2857
                                actual  : 14.285714285714286
                Test:           TestSla/MultipleStateChangesDecimalsFractionSeventhPart/Host
                Messages:       unexpected SLA value
=== RUN   TestSla/MultipleStateChangesDecimalsFractionSeventhPart/Service
    sla_test.go:310:
                Error Trace:    /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:310
                                                        /home/apenning/git/github.com/Icinga/icingadb/tests/sql/sla_test.go:277
                Error:          Not equal:
                                expected: 14.2857
                                actual  : 14.285714285714286
                Test:           TestSla/MultipleStateChangesDecimalsFractionSeventhPart/Service
                Messages:       unexpected SLA value
--- FAIL: TestSla (2.57s)
    --- PASS: TestSla/MultipleStateChangesDecimalsOddNumbers (0.02s)
        --- PASS: TestSla/MultipleStateChangesDecimalsOddNumbers/Host (0.01s)
        --- PASS: TestSla/MultipleStateChangesDecimalsOddNumbers/Service (0.01s)
    --- FAIL: TestSla/MultipleStateChangesDecimalsFractionOneThird (0.00s)
        --- FAIL: TestSla/MultipleStateChangesDecimalsFractionOneThird/Host (0.00s)
        --- FAIL: TestSla/MultipleStateChangesDecimalsFractionOneThird/Service (0.00s)
    --- FAIL: TestSla/MultipleStateChangesDecimalsFractionSeventhPart (0.00s)
        --- FAIL: TestSla/MultipleStateChangesDecimalsFractionSeventhPart/Host (0.00s)
        --- FAIL: TestSla/MultipleStateChangesDecimalsFractionSeventhPart/Service (0.00s)
FAIL
```

Reply:
To be more specific: of course this makes sense, but I did not notice it because my previous test worked with too kind IEEE 754 numbers. In the MySQL implementation, this error did not occur because the return value was stored in the `result` variable of type `decimal(7, 4)`.

Reply:
> To be more specific: of course this makes sense

Does it? I mean the function is defined as `RETURNS decimal(7, 4)`. Is this just a nonbinding then? :scream:

Reply:
> Is this just a nonbinding then? 😱

It depends, I guess. I played around a bit and it seems like everything numeric - all integers and floats - and `NULL` are valid for this return type. However, a string, for example, was not allowed. By accident I found [this post](https://dba.stackexchange.com/questions/280182/should-postgres-function-return-coerce-numeric-scale-precision) where someone experienced a similar behavior.